functions/go/set-image/transformer/images_transformer.go

```diff
@@ -0,0 +1,199 @@
+		}
+	}
+	return nil
+}
+
+// validateInput validates the inputs passed into via the functionConfig
+func (t *SetImage) validateInput() error {
+	// TODO: support container name and only one argument input in the next PR
+	if t.Image.Name == "" {
+		return fmt.Errorf("must specify `name`")
+	}
+	if t.Image.NewName == "" && t.Image.NewTag == "" && t.Image.Digest == "" {
+		return fmt.Errorf("must specify one of `newName`, `newTag`, or `digest`")
+	}
+	if t.Image.NewTag != "" && t.Image.Digest != "" {
+		return fmt.Errorf("image `newTag` and `digest` both set, set only one")
+	}
+	return nil
+}
+
+// updateContainerImages updates the images inside containers, return potential error, and a list of logging results
+func (t *SetImage) updateContainerImages(pod *fn.SubObject) (error, []setImageResult) {
+	var containers fn.SliceSubObjects
+	containers = append(containers, pod.GetSlice("iniContainers")...)
+	containers = append(containers, pod.GetSlice("containers")...)
+
+	var result []setImageResult
+	for _, o := range containers {
+		oldValue := o.NestedStringOrDie("image")
+		if !image.IsImageMatched(oldValue, t.Image.Name) {
+			continue
+		}
+		newName := getNewImageName(oldValue, t.Image)
+		if oldValue == newName {
+			return nil, nil
+		}
+
+		if err := o.SetNestedString(newName, "image"); err != nil {
+			return err, nil
+		}
+		t.resultCount += 1
+
+		result = append(result, setImageResult{
+			currentValue:  oldValue,
+			proposedValue: newName,
+		})
+	}
+	return nil, result
+}
+
+func (t *SetImage) setPodSpecContainers(o *fn.KubeObject) (error, []setImageResult) {
+	spec := o.GetMap("spec")
+	if spec == nil {
+		return nil, nil
+	}
+	template := spec.GetMap("template")
+	if template == nil {
+		return nil, nil
+	}
+	podSpec := template.GetMap("spec")
+	err, result := t.updateContainerImages(podSpec)
+	if err != nil {
+		return err, nil
+	}
+	return nil, result
+}
+
+func (t *SetImage) setPodContainers(o *fn.KubeObject) (error, []setImageResult) {
+	spec := o.GetMap("spec")
+	if spec == nil {
+		return nil, nil
+	}
+	err, result := t.updateContainerImages(spec)
+	if err != nil {
+		return err, nil
+	}
+	return nil, result
+}
+
+// getNewImageName return the new name for image field
+func getNewImageName(oldValue string, newImage Image) string {
+	name, tag, digest := image.Split(oldValue)
+	if newImage.NewName != "" {
+		name = newImage.NewName
+	}
+	if newImage.NewTag != "" {
+		tag = ":" + newImage.NewTag
+	}
+	if newImage.Digest != "" {
+		tag = "@" + newImage.Digest
+	}
+	var newName string
+	if tag == "" {
+		newName = name + digest
+	} else {
+		newName = name + tag
+	}
+
+	return newName
+}
+
+func (t SetImage) logResult(ctx *fn.Context, err error, result []setImageResult, o *fn.KubeObject) {
+	if err != nil {
+		ctx.ResultErr(err.Error(), o)
+	}
+	for _, val := range result {
+		msg := fmt.Sprintf("updated image from %v to %v", val.currentValue, val.proposedValue)
+		ctx.ResultInfo(msg, o)
+	}
```

Comment:
I might be a little bit out of sync about this. I thought we decide to not include each individual info updates in the `resourcelist.results`， right?

Reply:
Sorry about the confusion. I did not remember we discussed that. Previous version did include this message. So I kept it. I will delete it, and only leave the error messages and total info message

Reply:
The main problem is that kpt normally truncate the result if it is too long. And if we keep the detailed info for each resource change, more useful information like the summary and error logs will be truncated.

Reply:
The same problems happens on set-labels and we should fix that as well.

Reply:
My bad. I thought truncation would only leave out the early log results, and the total logging would remain.